### ISSUE TITLE
Missing @impl warning when using __using__ and overriding

### ISSUE THREAD

**Issue by asummers** [elixir-lang/elixir#7707]
### Environment

```
Erlang/OTP 20 [erts-9.3] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Elixir 1.6.5 (compiled with OTP 20)
```

### Current behavior
```
defmodule ExampleBehaviour do
  @callback foo() :: :ok
  @callback bar() :: :ok

  defmacro __using__(_) do
    quote do
      @impl ExampleBehaviour
      def foo(), do: :ok

      @impl ExampleBehaviour
      def bar(), do: :ok

      defoverridable foo: 0, bar: 0
    end
  end
end

defmodule WorkingExample do
  @behaviour ExampleBehaviour

  @impl ExampleBehaviour
  def foo(), do: :ok

  # This warns correctly about missing @impl
  def bar(), do: :ok
end

defmodule MissingWarningExample do
  @behaviour ExampleBehaviour

  use ExampleBehaviour

  # This does not warn about missing `@impl` despite the `__using__`
  # setting on `foo/0`.
  def bar(), do: :ok
end
```

### Expected behavior

When setting `@impl` for one function in a callback, Elixir warns if there's any missing `@impl`s (great feature by the way). Since the `__using__` macro sets `@impl` for a non zero number of the functions, dropping those functions into my module with the `__using__` macro should trigger the missing warnings on `bar/0`, I believe.

**Comment by asummers** [comment]
That behavior has definitely caused bugs in our code. We had one function spelled incorrectly that was overridden and a @impl would have shed light on that. Is that something you'd be interested in revisiting in a 2.x release, where by default you'd yell at missing @impls?

**Comment by josevalim** [comment]
Right. The best way to make sure you are overriding something is by always defining `@impl`. The behaviour of getting a warning if you forget an `@impl` is a convenience that `@impl` offers.

We haven't decided yet if we will always check for `@impl` on Elixir v2.0 but it will definitely be one of the topics for discussion.

**Comment by zachdaniel** [comment]
Would there be any kind of interest in like a sort of strict mode for a macro? To show all (or some of) the warnings that are suppressed normally?

**Comment by josevalim** [comment]
@zachdaniel we haven't thought about it although I don't think such mode would make a difference here. We do warn on macro code too, it is just that we consider each `@impl` based on its context.
